File: agenkit-ts/src/evaluation/prompt-optimizer.ts
/**
 * Prompt Optimization Framework
 *
 * Automatically improve prompts through systematic variation and testing
 * using grid search, random search, or genetic algorithms.
 *
 * Example:
 * ```typescript
 * const template = `You are a {role}.
 * {instructions}`;
 *
 * const variations = {
 *   role: ['helpful assistant', 'expert advisor'],
 *   instructions: ['Be concise.', 'Be detailed.'],
 * };
 *
 * const optimizer = new PromptOptimizer({
 *   template,
 *   variations,
 *   agentFactory: (prompt) => new MyAgent({ systemPrompt: prompt }),
 *   metrics: ['accuracy'],
 * });
 *
 * const result = await optimizer.optimize(testCases, 'grid');
 * console.log(`Best prompt: ${result.bestPrompt}`);
 * console.log(`Best scores: ${JSON.stringify(result.bestScores)}`);
 * ```
 */

import { Agent } from '../core/interfaces';
import { Evaluator, TestCase } from './core';
import { AccuracyMetric } from './quality-metrics';

/**
 * Optimization strategy types.
 */
export type OptimizationStrategy = 'grid' | 'random' | 'genetic';

/**
 * Agent factory that creates agents from prompt strings.
 */
export type PromptAgentFactory = (prompt: string) => Agent;

/**
 * Single prompt evaluation result.
 */
export interface PromptEvaluation {
  /** Generated prompt text */
  prompt: string;
  /** Variable configuration used */
  config: Record<string, string>;
  /** Metric scores achieved */
  scores: Record<string, number>;
}

/**
 * Results from prompt optimization.
 */
export interface PromptOptimizationResult {
  /** Best prompt found */
  bestPrompt: string;
  /** Best variable configuration */
  bestConfig: Record<string, string>;
  /** Best metric scores */
  bestScores: Record<string, number>;
  /** All evaluations performed */
  history: PromptEvaluation[];
  /** Number of prompts evaluated */
  nEvaluated: number;
  /** Strategy used */
  strategy: string;
  /** Optimization start time */
  startTime: Date;
  /** Optimization end time */
  endTime: Date;
}

/**
 * Get duration of prompt optimization in seconds.
 *
 * @param result Optimization result
 * @returns Duration in seconds
 */
export function getPromptOptimizationDuration(result: PromptOptimizationResult): number {
  return (result.endTime.getTime() - result.startTime.getTime()) / 1000;
}

/**
 * Convert prompt optimization result to plain object.
 *
 * @param result Optimization result
 * @returns Plain object representation
 */
export function promptOptimizationResultToDict(
  result: PromptOptimizationResult
): Record<string, unknown> {
  return {
    best_prompt: result.bestPrompt,
    best_config: result.bestConfig,
    best_scores: result.bestScores,
    n_evaluated: result.nEvaluated,
    strategy: result.strategy,
    duration_seconds: getPromptOptimizationDuration(result),
    start_time: result.startTime.toISOString(),
    end_time: result.endTime.toISOString(),
  };
}

/**
 * Configuration for prompt optimizer.
 */
export interface PromptOptimizerConfig {
  /** Prompt template with {variable} placeholders */
  template: string;
  /** Map of variable names to possible values */
  variations: Record<string, string[]>;
  /** Function that creates agent from prompt string */
  agentFactory: PromptAgentFactory;
  /** List of metrics to evaluate */
  metrics: string[];
  /** Primary metric for optimization (defaults to first metric) */
  objectiveMetric?: string;
  /** Whether to maximize (true) or minimize (false) objective */
  maximize?: boolean;
}

/**
 * Genetic algorithm configuration.
 */
export interface GeneticConfig {
  /** Population size */
  populationSize?: number;
  /** Number of generations */
  nGenerations?: number;
  /** Mutation rate (0-1) */
  mutationRate?: number;
}

/**
 * Optimize prompts through systematic variation.
 *
 * Supports multiple optimization strategies:
 * - Grid search: Exhaustive evaluation of all combinations
 * - Random search: Random sampling of combinations
 * - Genetic algorithm: Evolutionary optimization
 */
export class PromptOptimizer {
  private template: string;
  private variations: Record<string, string[]>;
  private agentFactory: PromptAgentFactory;
  private metrics: string[];
  private objectiveMetric: string;
  private maximize: boolean;
  private history: PromptEvaluation[] = [];

  /**
   * Create prompt optimizer.
   *
   * @param config Optimizer configuration
   */
  constructor(config: PromptOptimizerConfig) {
    this.template = config.template;
    this.variations = config.variations;
    this.agentFactory = config.agentFactory;
    this.metrics = config.metrics;
    this.objectiveMetric = config.objectiveMetric || config.metrics[0];
    this.maximize = config.maximize ?? true;
  }

  /**
   * Optimize prompt using specified strategy.
   *
   * @param testCases Test cases for evaluation
   * @param strategy Optimization strategy
   * @param options Strategy-specific options
   * @returns Optimization result
   */
  async optimize(
    testCases: TestCase[],
    strategy: OptimizationStrategy,
    options?: Record<string, unknown>
  ): Promise<PromptOptimizationResult> {
    const startTime = new Date();
    this.history = [];

    let result: PromptOptimizationResult;

    switch (strategy) {
      case 'grid':
        result = await this.optimizeGrid(testCases);
        break;
      case 'random':
        const nSamples = (options?.nSamples as number) || 20;
        result = await this.optimizeRandom(testCases, nSamples);
        break;
      case 'genetic':
        const geneticConfig: GeneticConfig = {
          populationSize: (options?.populationSize as number) || 10,
          nGenerations: (options?.nGenerations as number) || 10,
          mutationRate: (options?.mutationRate as number) || 0.2,
        };
        result = await this.optimizeGenetic(testCases, geneticConfig);
        break;
      default:
        throw new Error(`Unknown strategy: ${strategy}`);
    }

    result.endTime = new Date();
    return result;
  }

  /**
   * Grid search: exhaustive evaluation of all combinations.
   *
   * @param testCases Test cases for evaluation
   * @returns Optimization result
   */
  private async optimizeGrid(testCases: TestCase[]): Promise<PromptOptimizationResult> {
    const configs = this.generateAllConfigs();

    let bestConfig: Record<string, string> | null = null;
    let bestScores: Record<string, number> = {};
    let bestObjective = this.maximize ? -Infinity : Infinity;

    for (const config of configs) {
      const prompt = this.fillTemplate(config);
      const scores = await this.evaluatePrompt(prompt, config, testCases);

      const objective = scores[this.objectiveMetric] || 0;
      const isBetter = this.maximize ? objective > bestObjective : objective < bestObjective;

      if (isBetter) {
        bestObjective = objective;
        bestConfig = config;
        bestScores = scores;
      }
    }

    return {
      bestPrompt: this.fillTemplate(bestConfig!),
      bestConfig: bestConfig!,
      bestScores,
      history: this.history,
      nEvaluated: configs.length,
      strategy: 'grid',
      startTime: new Date(),
      endTime: new Date(),
    };
  }

  /**
   * Random search: sample random combinations.
   *
   * @param testCases Test cases for evaluation
   * @param nSamples Number of samples to evaluate
   * @returns Optimization result
   */
  private async optimizeRandom(
    testCases: TestCase[],
    nSamples: number
  ): Promise<PromptOptimizationResult> {
    let bestConfig: Record<string, string> | null = null;
    let bestScores: Record<string, number> = {};
    let bestObjective = this.maximize ? -Infinity : Infinity;

    for (let i = 0; i < nSamples; i++) {
      const config = this.sampleConfig();
      const prompt = this.fillTemplate(config);
      const scores = await this.evaluatePrompt(prompt, config, testCases);

      const objective = scores[this.objectiveMetric] || 0;
      const isBetter = this.maximize ? objective > bestObjective : objective < bestObjective;

      if (isBetter) {
        bestObjective = objective;
        bestConfig = config;
        bestScores = scores;
      }
    }

    return {
      bestPrompt: this.fillTemplate(bestConfig!),
      bestConfig: bestConfig!,
      bestScores,
      history: this.history,
      nEvaluated: nSamples,
      strategy: 'random',
      startTime: new Date(),
      endTime: new Date(),
    };
  }

  /**
   * Genetic algorithm: evolutionary optimization.
   *
   * @param testCases Test cases for evaluation
   * @param config Genetic algorithm configuration
   * @returns Optimization result
   */
  private async optimizeGenetic(
    testCases: TestCase[],
    config: GeneticConfig
  ): Promise<PromptOptimizationResult> {
    const populationSize = config.populationSize || 10;
    const nGenerations = config.nGenerations || 10;
    const mutationRate = config.mutationRate || 0.2;

    // Initialize population
    let population: Array<{ config: Record<string, string>; fitness: number }> = [];
    for (let i = 0; i < populationSize; i++) {
      const cfg = this.sampleConfig();
      const prompt = this.fillTemplate(cfg);
      const scores = await this.evaluatePrompt(prompt, cfg, testCases);
      const fitness = scores[this.objectiveMetric] || 0;
      population.push({ config: cfg, fitness });
    }

    // Evolve
    for (let gen = 0; gen < nGenerations; gen++) {
      // Sort by fitness
      population.sort((a, b) => (this.maximize ? b.fitness - a.fitness : a.fitness - b.fitness));

      // Select top half (elitism)
      const survivors = population.slice(0, Math.floor(populationSize / 2));

      // Create offspring
      const offspring: typeof population = [];
      while (offspring.length + survivors.length < populationSize) {
        // Tournament selection
        const parent1 = this.tournamentSelect(survivors);
        const parent2 = this.tournamentSelect(survivors);

        // Crossover
        const child = this.crossover(parent1.config, parent2.config);

        // Mutation
        if (Math.random() < mutationRate) {
          this.mutate(child);
        }

        // Evaluate
        const prompt = this.fillTemplate(child);
        const scores = await this.evaluatePrompt(prompt, child, testCases);
        const fitness = scores[this.objectiveMetric] || 0;
        offspring.push({ config: child, fitness });
      }

      population = [...survivors, ...offspring];
    }

    // Find best
    population.sort((a, b) => (this.maximize ? b.fitness - a.fitness : a.fitness - b.fitness));
    const best = population[0];

    return {
      bestPrompt: this.fillTemplate(best.config),
      bestConfig: best.config,
      bestScores: { [this.objectiveMetric]: best.fitness },
      history: this.history,
      nEvaluated: this.history.length,
      strategy: 'genetic',
      startTime: new Date(),
      endTime: new Date(),
    };
  }

  /**
   * Fill template with configuration values.
   *
   * @param config Variable configuration
   * @returns Filled prompt
   */
  private fillTemplate(config: Record<string, string>): string {
    let result = this.template;
    for (const [key, value] of Object.entries(config)) {
      const placeholder = `{${key}}`;
      result = result.replace(new RegExp(placeholder, 'g'), value);
    }
    return result;
  }

  /**
   * Generate all possible configurations (Cartesian product).
   *
   * @returns All configuration combinations
   */
  private generateAllConfigs(): Array<Record<string, string>> {
    const keys = Object.keys(this.variations);
    const valueLists = keys.map(key => this.variations[key]);

    return this.cartesianProduct(keys, valueLists, 0, {});
  }

  /**
   * Generate Cartesian product recursively.
   *
   * @param keys Variable names
   * @param valueLists Lists of values for each variable
   * @param index Current index
   * @param current Current configuration
   * @returns All configurations
   */
  private cartesianProduct(
    keys: string[],
    valueLists: string[][],
    index: number,
    current: Record<string, string>
  ): Array<Record<string, string>> {
    if (index === keys.length) {
      return [{ ...current }];
    }

    const results: Array<Record<string, string>> = [];
    for (const value of valueLists[index]) {
      current[keys[index]] = value;
      const configs = this.cartesianProduct(keys, valueLists, index + 1, current);
      results.push(...configs);
    }

    return results;
  }

  /**
   * Sample random configuration.
   *
   * @returns Random configuration
   */
  private sampleConfig(): Record<string, string> {
    const config: Record<string, string> = {};
    for (const [key, values] of Object.entries(this.variations)) {
      config[key] = values[Math.floor(Math.random() * values.length)];
    }
    return config;
  }

  /**
   * Evaluate prompt on test cases.
   *
   * @param prompt Prompt to evaluate
   * @param config Configuration used
   * @param testCases Test cases
   * @returns Metric scores
   */
  private async evaluatePrompt(
    prompt: string,
    config: Record<string, string>,
    testCases: TestCase[]
  ): Promise<Record<string, number>> {
    const agent = this.agentFactory(prompt);

    // Create metrics
    const metricObjects = [];
    if (this.metrics.includes('accuracy')) {
      metricObjects.push(new AccuracyMetric());
    }

    const evaluator = new Evaluator(agent, metricObjects);
    const result = await evaluator.evaluate(testCases);

    // Extract scores
    const scores: Record<string, number> = {};
    for (const metric of this.metrics) {
      if (metric === 'accuracy' && result.accuracy !== undefined) {
        scores.accuracy = result.accuracy;
      } else if (metric === 'quality_score' && result.qualityScore !== undefined) {
        scores.quality_score = result.qualityScore;
      } else if (metric === 'latency_ms' && result.avgLatencyMs !== undefined) {
        scores.latency_ms = result.avgLatencyMs;
      }
    }

    // Record evaluation
    this.history.push({ prompt, config, scores });

    return scores;
  }

  /**
   * Tournament selection for genetic algorithm.
   *
   * @param population Population to select from
   * @returns Selected individual
   */
  private tournamentSelect(
    population: Array<{ config: Record<string, string>; fitness: number }>
  ): { config: Record<string, string>; fitness: number } {
    const tournamentSize = 3;
    const tournament = [];

    for (let i = 0; i < tournamentSize; i++) {
      const idx = Math.floor(Math.random() * population.length);
      tournament.push(population[idx]);
    }

    tournament.sort((a, b) => (this.maximize ? b.fitness - a.fitness : a.fitness - b.fitness));
    return tournament[0];
  }

  /**
   * Crossover two configurations.
   *
   * @param parent1 First parent
   * @param parent2 Second parent
   * @returns Child configuration
   */
  private crossover(
    parent1: Record<string, string>,
    parent2: Record<string, string>
  ): Record<string, string> {
    const child: Record<string, string> = {};
    for (const key of Object.keys(this.variations)) {
      child[key] = Math.random() < 0.5 ? parent1[key] : parent2[key];
    }
    return child;
  }

  /**
   * Mutate configuration.
   *
   * @param config Configuration to mutate (mutated in place)
   */
  private mutate(config: Record<string, string>): void {
    const keys = Object.keys(this.variations);
    const keyToMutate = keys[Math.floor(Math.random() * keys.length)];
    const values = this.variations[keyToMutate];
    config[keyToMutate] = values[Math.floor(Math.random() * values.length)];
  }

  /**
   * Get optimization history.
   *
   * @returns History of prompt evaluations
   */
  getHistory(): PromptEvaluation[] {
    return [...this.history];
  }
}

File: agenkit/evaluation/prompt_optimizer.py
"""
Prompt Optimization Framework

Automatically improve prompts through systematic variation and testing
using grid search, random search, or genetic algorithms.

Example:
    >>> from agenkit.evaluation import PromptOptimizer
    >>>
    >>> template = '''
    ... You are a {role}.
    ... {instructions}
    ... '''
    >>>
    >>> variations = {
    ...     "role": ["helpful assistant", "expert advisor"],
    ...     "instructions": ["Be concise.", "Be detailed."]
    ... }
    >>>
    >>> optimizer = PromptOptimizer(
    ...     template=template,
    ...     variations=variations,
    ...     agent_factory=lambda prompt: MyAgent(system_prompt=prompt),
    ...     metrics=["accuracy"]
    ... )
    >>>
    >>> result = await optimizer.optimize(test_cases)
"""

import random
from collections.abc import Callable
from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum
from typing import Any

from .core import Evaluator


class OptimizationStrategy(Enum):
    """Prompt optimization strategies."""

    GRID = "grid"  # Exhaustive grid search
    RANDOM = "random"  # Random sampling
    GENETIC = "genetic"  # Genetic algorithm


@dataclass
class PromptOptimizationResult:
    """
    Results from prompt optimization.

    Attributes:
        best_prompt: Best prompt found
        best_config: Best variable configuration
        best_scores: Best metric scores
        history: List of (prompt, config, scores) tuples
        n_evaluated: Number of prompts evaluated
        strategy: Optimization strategy used
        start_time: Optimization start timestamp
        end_time: Optimization end timestamp
    """

    best_prompt: str
    best_config: dict[str, str]
    best_scores: dict[str, float]
    history: list[tuple[str, dict[str, str], dict[str, float]]]
    n_evaluated: int
    strategy: str
    start_time: str
    end_time: str

    @property
    def duration_seconds(self) -> float:
        """Duration of optimization in seconds."""
        start = datetime.fromisoformat(self.start_time)
        end = datetime.fromisoformat(self.end_time)
        return (end - start).total_seconds()

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "best_prompt": self.best_prompt,
            "best_config": self.best_config,
            "best_scores": self.best_scores,
            "n_evaluated": self.n_evaluated,
            "strategy": self.strategy,
            "duration_seconds": self.duration_seconds,
            "start_time": self.start_time,
            "end_time": self.end_time,
        }


class PromptOptimizer:
    """
    Prompt optimization through systematic variation.

    Supports multiple optimization strategies:
    - Grid search: Exhaustive evaluation of all combinations
    - Random search: Random sampling of combinations
    - Genetic algorithm: Evolutionary optimization

    Example:
        >>> optimizer = PromptOptimizer(
        ...     template="You are a {role}. {instructions}",
        ...     variations={
        ...         "role": ["assistant", "advisor", "guide"],
        ...         "instructions": ["Be brief.", "Be detailed."]
        ...     },
        ...     agent_factory=lambda p: MyAgent(system_prompt=p),
        ...     metrics=["accuracy", "quality_score"]
        ... )
        >>> result = await optimizer.optimize(test_cases, strategy="grid")
    """

    def __init__(
        self,
        template: str,
        variations: dict[str, list[str]],
        agent_factory: Callable[[str], Any],
        metrics: list[str],
        objective_metric: str | None = None,
        maximize: bool = True,
    ):
        """
        Initialize prompt optimizer.

        Args:
            template: Prompt template with {variable} placeholders
            variations: Dict mapping variable names to possible values
            agent_factory: Function that creates agent from prompt string
            metrics: List of metrics to evaluate
            objective_metric: Primary metric for optimization (default: first metric)
            maximize: Whether to maximize (True) or minimize (False) objective
        """
        self.template = template
        self.variations = variations
        self.agent_factory = agent_factory
        self.metrics = metrics
        self.objective_metric = objective_metric or metrics[0]
        self.maximize = maximize
        self.history: list[tuple[str, dict[str, str], dict[str, float]]] = []

    def _fill_template(self, config: dict[str, str]) -> str:
        """Fill template with configuration values."""
        return self.template.format(**config)

    def _generate_all_configs(self) -> list[dict[str, str]]:
        """Generate all possible configurations (Cartesian product)."""
        import itertools

        keys = list(self.variations.keys())
        values = [self.variations[k] for k in keys]

        configs = []
        for combo in itertools.product(*values):
            config = dict(zip(keys, combo, strict=False))
            configs.append(config)

        return configs

    def _sample_config(self) -> dict[str, str]:
        """Sample random configuration."""
        return {key: random.choice(values) for key, values in self.variations.items()}

    async def _evaluate_prompt(self, prompt: str, test_cases: list[dict[str, Any]]) -> dict[str, float]:
        """
        Evaluate prompt on test cases.

        Args:
            prompt: Prompt to evaluate
            test_cases: Test cases for evaluation

        Returns:
            Dict of metric scores
        """
        # Create agent with prompt
        agent = self.agent_factory(prompt)

        # Evaluate
        from agenkit.evaluation.quality_metrics import AccuracyMetric, QualityMetrics

        # Create metrics
        metric_objs = []
        if "accuracy" in self.metrics:
            metric_objs.append(AccuracyMetric(None, False))
        if "quality_score" in self.metrics:
            metric_objs.append(QualityMetrics(False, "", None))

        evaluator = Evaluator(agent, metric_objs, "prompt-opt")
        result = await evaluator.evaluate(test_cases, "")

        # Extract scores
        scores = {}
        if "accuracy" in self.metrics and result.accuracy is not None:
            scores["accuracy"] = result.accuracy
        if "quality_score" in self.metrics and result.quality_score is not None:
            scores["quality_score"] = result.quality_score
        if "latency_ms" in self.metrics and result.avg_latency_ms is not None:
            scores["latency_ms"] = result.avg_latency_ms

        return scores

    def _get_objective_score(self, scores: dict[str, float]) -> float:
        """Get objective score from metric scores."""
        score = scores.get(self.objective_metric, 0.0)

        # Invert if minimizing (e.g., latency)
        if not self.maximize:
            score = -score

        return score

    async def optimize_grid(
        self, test_cases: list[dict[str, Any]]
    ) -> PromptOptimizationResult:
        """
        Grid search: Evaluate all possible combinations.

        Args:
            test_cases: Test cases for evaluation

        Returns:
            PromptOptimizationResult
        """
        start_time = datetime.now(timezone.utc).isoformat()
        self.history = []

        # Generate all configs
        configs = self._generate_all_configs()

        best_prompt = ""
        best_config: dict[str, str] = {}
        best_scores: dict[str, float] = {}
        best_objective = float("-inf")

        # Evaluate each configuration
        for config in configs:
            prompt = self._fill_template(config)
            scores = await self._evaluate_prompt(prompt, test_cases)
            objective_score = self._get_objective_score(scores)

            self.history.append((prompt, config.copy(), scores.copy()))

            if objective_score > best_objective:
                best_objective = objective_score
                best_prompt = prompt
                best_config = config.copy()
                best_scores = scores.copy()

        end_time = datetime.now(timezone.utc).isoformat()

        return PromptOptimizationResult(
            best_prompt=best_prompt,
            best_config=best_config,
            best_scores=best_scores,
            history=self.history,
            n_evaluated=len(configs),
            strategy="grid",
            start_time=start_time,
            end_time=end_time,
        )

    async def optimize_random(
        self, test_cases: list[dict[str, Any]], n_samples: int = 20
    ) -> PromptOptimizationResult:
        """
        Random search: Sample random combinations.

        Args:
            test_cases: Test cases for evaluation
            n_samples: Number of random samples to evaluate

        Returns:
            PromptOptimizationResult
        """
        start_time = datetime.now(timezone.utc).isoformat()
        self.history = []

        best_prompt = ""
        best_config: dict[str, str] = {}
        best_scores: dict[str, float] = {}
        best_objective = float("-inf")

        # Sample and evaluate random configurations
        for _ in range(n_samples):
            config = self._sample_config()
            prompt = self._fill_template(config)
            scores = await self._evaluate_prompt(prompt, test_cases)
            objective_score = self._get_objective_score(scores)

            self.history.append((prompt, config.copy(), scores.copy()))

            if objective_score > best_objective:
                best_objective = objective_score
                best_prompt = prompt
                best_config = config.copy()
                best_scores = scores.copy()

        end_time = datetime.now(timezone.utc).isoformat()

        return PromptOptimizationResult(
            best_prompt=best_prompt,
            best_config=best_config,
            best_scores=best_scores,
            history=self.history,
            n_evaluated=n_samples,
            strategy="random",
            start_time=start_time,
            end_time=end_time,
        )

    async def optimize_genetic(
        self,
        test_cases: list[dict[str, Any]],
        population_size: int = 10,
        n_generations: int = 5,
        mutation_rate: float = 0.2,
    ) -> PromptOptimizationResult:
        """
        Genetic algorithm: Evolve prompts through selection and mutation.

        Args:
            test_cases: Test cases for evaluation
            population_size: Size of population
            n_generations: Number of generations
            mutation_rate: Probability of mutation per gene

        Returns:
            PromptOptimizationResult
        """
        start_time = datetime.now(timezone.utc).isoformat()
        self.history = []

        # Initialize population with random configurations
        population = [self._sample_config() for _ in range(population_size)]
        fitness_scores: list[float] = []

        # Evaluate initial population
        for config in population:
            prompt = self._fill_template(config)
            scores = await self._evaluate_prompt(prompt, test_cases)
            objective_score = self._get_objective_score(scores)
            fitness_scores.append(objective_score)
            self.history.append((prompt, config.copy(), scores.copy()))

        # Evolution loop
        for _ in range(n_generations):
            # Selection: Tournament selection
            new_population = []
            for _ in range(population_size):
                # Select 2 random individuals
                idx1, idx2 = random.sample(range(population_size), 2)
                # Choose fitter one
                winner_idx = idx1 if fitness_scores[idx1] > fitness_scores[idx2] else idx2
                new_population.append(population[winner_idx].copy())

            # Mutation
            for config in new_population:
                for key in config:
                    if random.random() < mutation_rate:
                        config[key] = random.choice(self.variations[key])

            # Evaluate new population
            population = new_population
            fitness_scores = []

            for config in population:
                prompt = self._fill_template(config)
                scores = await self._evaluate_prompt(prompt, test_cases)
                objective_score = self._get_objective_score(scores)
                fitness_scores.append(objective_score)
                self.history.append((prompt, config.copy(), scores.copy()))

        # Find best from all history
        best_idx = max(range(len(self.history)), key=lambda i: self._get_objective_score(self.history[i][2]))
        best_prompt, best_config, best_scores = self.history[best_idx]

        end_time = datetime.now(timezone.utc).isoformat()

        return PromptOptimizationResult(
            best_prompt=best_prompt,
            best_config=best_config,
            best_scores=best_scores,
            history=self.history,
            n_evaluated=len(self.history),
            strategy="genetic",
            start_time=start_time,
            end_time=end_time,
        )

    async def optimize(
        self,
        test_cases: list[dict[str, Any]],
        strategy: str = "random",
        **kwargs: Any,
    ) -> PromptOptimizationResult:
        """
        Run prompt optimization.

        Args:
            test_cases: Test cases for evaluation
            strategy: Optimization strategy ("grid", "random", "genetic")
            **kwargs: Strategy-specific parameters

        Returns:
            PromptOptimizationResult
        """
        if strategy == "grid":
            return await self.optimize_grid(test_cases)
        if strategy == "random":
            n_samples = kwargs.get("n_samples", 20)
            return await self.optimize_random(test_cases, n_samples=n_samples)
        if strategy == "genetic":
            population_size = kwargs.get("population_size", 10)
            n_generations = kwargs.get("n_generations", 5)
            mutation_rate = kwargs.get("mutation_rate", 0.2)
            return await self.optimize_genetic(
                test_cases,
                population_size=population_size,
                n_generations=n_generations,
                mutation_rate=mutation_rate,
            )
        msg = f"Unknown strategy: {strategy}"
        raise ValueError(msg)
